content: draft: Harden 'safe-expunging-process' (#1203)
slsa-framework/slsa: docs/spec/draft/source-requirements.md

## Safe Expunging Process

SCSs MAY allow the organization to expunge (remove) content from a repository and its change history without leaving a public record of the removed content,
but MUST only allow these changes in order to meet legal or privacy compliance requirements.
Content changed under this process includes changing files, history, references, or any other metadata stored by the SCS.

### Warning

Removing a revision from a repository is similar to deleting a package version from a registry: it's almost impossible to estimate the amount of downstream supply chain impact.
For example, in VCSs like Git, removal of a revision changes the object IDs of all subsequent revisions that were built on top of it,

> this can break downstream consumers ability to refer to source they've already integrated into their products.

Reviewer: ```suggestion
and this can break downstream consumers ability to refer to source they've already integrated into their products.
```

nit

Author: So I was really trying to explain why it's bad that the object ID changes.

So maybe something more like:
```suggestion
, breaking downstream consumers ability to refer to source they've already integrated into their products.
```

Reviewer: Looks good. Follow up nit I just noticed: `breaking downstream consumers' ability to`, with the possessive apostrophe.

> It may be the case that the specific set of commits targeted by the takedown can be expunged in ways that do not impact revisions which can mitigate these problems.

Reviewer: ```suggestion
It may be the case that the specific set of changes targeted by the takedown can be expunged in ways that do not impact consumed revisions, which can mitigate these problems.
```

nit: making language a bit more generic and highlighting that "revisions" is the consumed ones from the previous sentence.

> It may be the case that the specific set of changes targeted by the takedown can be expunged in ways that do not impact consumed revisions, which can mitigate these problems.

Reviewer 2: Nit: I don't think we've used the term takedown anywhere before, so adding this for clarity

```suggestion
It may be the case that the specific set of changes targeted by a legal takedown can be expunged in ways that do not impact consumed revisions, which can mitigate these problems.
```

It is also the case that removing content from a repository won't necessarily remove it everywhere.
The content may still exist in other copies of the repository, either in backups or on developer machines.

### Process

An SCS MUST document the Safe Expunging Process and describe how requests and actions are tracked and SHOULD log the fact that content was removed.
Different organizations and tech stacks may have different approaches to the problem.

> SCSs SHOULD require multi-party approval for any changes made under this process.

Reviewer 2: This requirement isn't entire clear. Who are these multiple parties? And how are they authorized to approve a legal takedown?

Author: Reworded this.  Hopefully the new words are better?

## Safe Expunging Process

SCSs MAY allow the organization to expunge (remove) content from a repository and its change history without leaving a public record of the removed content.
This includes changing files, history, or changing references in git and is used to accommodate legal/privacy compliance requirements.

Removing a revision from a repository is similar to deleting a package version from a registry: it's almost impossible to estimate the amount of downstream supply chain impact.

> In version control systems like git, removal of a revision changes the object id of all subsequent revisions that were built on top of it.

Reviewer: ```suggestion
In version control systems like Git, removal of a revision changes the object IDs of all subsequent revisions that were built on top of it.
```

> Although there is no "safe" way to do it, it sometimes necessary and there are steps you can take to mitigate the damage.

Reviewer: ```suggestion
Although there is no "safe" way to do it, it is sometimes necessary and there are steps you can take to mitigate the damage.
```

Reviewer: I think it might flow better to swap this sentence with the previous one, especially since the "it" here is different from the "it" the last sentence ends with.

Reviewer 3: I think we need to be clearer also with no "safe" way. I understand we're using quotation marks around safe to indicate irony but I think it is easy for it to get confused especially since above we call it safe expunging process.

Maybe switching to something like

> Although there is no completely safe way to do it

Or we if we think even that is confusing maybe we call it something other than safe expunging process. Given that this is part of a specification it will be potentially interpreted in ways we can't anticipate, so anything we can do to clarify would be great.

> When used as an attack, this is called “repo hijacking” (or “repo-jacking”) and is one of the primary threats source provenance attestations protect against.

Reviewer: Does this text still belong here? I always understood this to mean repo takeovers or name squatting if a popular repository is renamed, and I think we've taken out repo renames from this.

As such great care must be taken by SCSs when implementing these changes for legitimate purposes.

SCSs SHOULD require multi-party approval for any changes made under this process.

The Safe Expunging Process MUST be documented and describe how requests and actions are tracked and SHOULD log the fact that content was removed.

On the git VCS, force pushes allow you to remove data from a branch.
If a branch has been identified as consumable branch, force pushes to that branch must follow the safe expunging process.

Different organizations and tech stacks may have different approaches to the problem.

Safe Expunging Scenarios:

> ### Legal Takedowns

Reviewer: If we have only one scenario, I wonder about reframing the section to say it's for legal takedowns only. OTOH, if we foresee adding scenarios, that'd create more work then, so we can punt until we're ready to publish and see if we still only have one scenario?

Author: I've reworked the whole thing and only cover legal takedowns.  There's now a section for 'Warning' where we tell folks the problems they're going to encounter and then a section for 'Process' where we talk a bit more about what to do.

The only currently acceptable scenario to expunge data is in response to some legal or compliance request.

> The takedown request MUST be addressed by following an agreed-upon process.

Reviewer: This agreed-upon process is the same as the documented safe expunging process mentioned above, right? If so, perhaps this could be combined or refer to it?

Author: Yeah, there's lots of duplication.  I've re-written a bunch and should have addressed this.

It may be the case that the specific set of commits targeted by the takedown can be expunged in ways that do not impact revisions.

> This process and the resulting logs MAY be private to both prevent calling attention to potentially sensitive data (e.g. PII) or to comply with local laws

Reviewer: ```suggestion
The application of the safe expunging process and the resulting logs MAY be private to both prevent calling attention to potentially sensitive data (e.g. PII) or to comply with local laws
```

Also, could this say something to the effect of "It must be as transparent / public as possible within the constraints of complying with local laws and regulations and protection of sensitive data (e.g. PII)." I think, as it's written, it's easy to default to making these actions fully private rather than a best effort at transparency?

Author: Honestly my preference is to accept your suggestion but leave it otherwise unchanged.  Setting up all the mechanisms for this will be very difficult.  If we say "must be as transparent/public as possible" we _will_ get questions like "how do we quantify that?", "what counts as transparent?", etc...

I'd prefer to err on the side of being less prescriptive here so that folks can more easily adopt.

Author: I did add something that says they SHOULD disclose if possible, which is hopefully easy enough?

## Safe Expunging Process

> SCSs MAY allow _administrators_ to expunge (remove) content from a repository and its change history without leaving a public record of the removed content.

Reviewer 4: ```suggestion
SCSs MAY allow the organization to expunge (remove) content from a repository and its change history without leaving a public record of the removed content.
```

I'm hoping we can remove this as a general capability of repo admins one day. It's logically going to be "the org" (not just the admins) that make decisions about which data is inappropriate for the repo and must be removed, because the org defines the "intent" of the code.

Author: agreed

This includes changing files, history, or changing references in git and is used to accommodate legal/privacy compliance requirements as well as administrative
changes within a repository (see below for more information on the various scenarios).
When used as an attack, this is called “repo hijacking” (or “repo-jacking”) and is one of the primary threats source provenance attestations protect against.
As such great care must be taken by SCSs when implementing these changes for legitimate purposes.

SCSs SHOULD require multi-party approval for any changes made under this process.

> The Safe Expunging Process MUST be documented and describe how requests and actions are tracked and SHOULD log the fact that content was removed.

Reviewer: Should they log _privately_ the content that was removed as well? is there an expectation of being able to _prove_ that only the content they were legally supposed to remove was removed, even if they can only do so for internal parties / auditors?

Reviewer: I guess this is hard to answer if the information that must be removed is something personal (say PII the person in question doesn't want a cloud SCP to hold on to even privately)...

Author: I think we're quite constrained here.  Different legal environments have different requirements.  If you think about [national security letters](https://en.wikipedia.org/wiki/National_security_letter) and their non-disclosure requirement, you could imagine a similar mechanism being used for a takedown request.  Of course, we need to think beyond just US law.  One could _imagine_ (I am not a lawyer!) [this Australian law](https://www.homeaffairs.gov.au/about-us/our-portfolios/national-security/lawful-access-telecommunications/assistance-and-access-overview) being used to require a change without logging it.

Additionally, requiring this logging mechanism would likely require a lot of new code and systems to be written, which would constrain implementers.  Maybe that's OK for SLSA L3, but it seems like a lot to ask for SLSA L2.  How readily do we want this to be adoptable?  How much do we want to delay the release and adoption of the source track while we get an answer to this question and implementations built?

Reviewer: Good points! I agree on not blocking too much on this and on not being very specific about how this is implemented as we can't foresee all scenarios and their variables.

The Safe Expunging Process MUST be documented and describe how requests and actions are tracked and SHOULD log the fact that content was removed.

On the git VCS, force pushes allow you to remove data from a branch.
If a branch has been identified as consumable branch, force pushes to that branch must follow the safe expunging process.

TODO: Determine how organizations can provide transparency around this process.
At a minimum the organization would need to declare why data was removed from the branch.

The goal of this sections is to document that this process is allowed.
Different organizations and tech stacks may have different approaches to the problem.

> Scenarios that need to be addressed:

Reviewer 4: this was definitely "things to think about" more than scenarios we know we want to support.

Author: I think we can resolve this since we've decided to only focus on legal requests for now.

> Safe Expunging Scenarios:

Reviewer: Should we highlight this list isn't exhaustive?

Author: Done!

> ### Legal Takedowns

Reviewer 4: In git / merkle-tree-based-VCS, I think this is going to be the only legit reason to consider history rewrites due to the damage they cause to the trustworthiness of a repo. 

In these cases, we should probably build out a health mapping concept. 
These shas XYZ have been updated to ABC. 

We need to somehow say: provenance issued for XYZ is legitimate and still hosted but the revision is no longer available in the repo. 
It will depend on the situation whether packages built on XYZ are still legal.

Author: Removed metadata/history rewrites entirely, so I think this comment is obsolete.

This process and the resulting logs MAY be private to both prevent calling attention to potentially sensitive data (e.g. PII) or to comply with local laws
and regulations which may require the change to be kept private to the extent possible.

### Commit metadata rewrites

A team may decide that all reachable commits in the history of a revision need to follow a new metadata convention.
In git VCS, compliance with this new policy will require history to be rewritten (commit metadata is included in the computation of the revision id).
Policies in this category include things like commit signatures, author / committer formatting restrictions, closed-issue-linkage, etc.

> SCSs SHOULD make the fact of these changes public when they occur (e.g. an issue, announcement, or other mechanism).

Reviewer: Thoughts on making this `SHOULD` a `MUST`? I understand why this can't be said for legal takedowns but this scenario (and others like it such as repo renames) seem to have no reason we can't require an announcement? If we can think of reasons, perhaps we list them as examples of when an announcement isn't made?

Author: Sure.  I also changed SCSs to 'Organizations' since it's likely not the system that will do this but rather the people involved in running it.

Reviewer: Thanks!

> Organizations MUST make the fact of these changes public when they occur (e.g. an issue, announcement, or other mechanism).

Reviewer 4: I've been struggling with this topic recently. 

I'm certainly leaning towards a position that "we wish commit metadata were different" should never be a good enough reason to rewrite history. 
Commit shas are the foreign keys of package registries. Changing them causes unknowable damage to an ecosystem's ability to trust itself.

Author: I can add some language to that effect.  "Orgs should strive to avoid rewriting commit metadata and only do so as a last resort" ?

Reviewer 4: I was considering just cutting this "commit metadata rewrites" section. Essentially saying "This process is built for legal challenges. Other uses are hard to justify."

Reviewer: I like the idea of reducing the scope of use of safe expunging to legal challenges.

Author: That works for me.  Would I remove 'Repository renames' too?

Reviewer 4: good question -- maybe we could move that info into "how source provenance attestation subjects work?"

Essentially the take away should be "if you remove a sha or change the url of a repo, provenance attestations are voided. Also you may have broken the chain of custody for shipped packages."

Author: Done

Reviewer 2: @TomHennen Is there an issue tracking the repo renaming discussion?

Author: No, I didn't create one.  It sounded as though we could drop it until we needed to resolve it.  Should I create one anyways?

> ### Repository renames

Reviewer 4: probably the best case scenario here is to just archive/lock the previous repo name so the attestations stay hosted and available.

When a repo is transferred to a new organization ("donated"), or if a repo must be renamed or otherwise have its url changed within the same org, attestations for previous revisions of this repo will no longer be matched because the combination of the repository id and the revision id will have changed.

> Organizations MUST make the fact of these changes public when they occur (e.g. an issue, announcement, or other mechanism).

Reviewer 4: I think this would only be true for public oss repos. 

```suggestion
Organizations MUST inform all consumers of a repo of these changes when they occur.
Organizations SHOULD provide a means to resolve previous revision IDs to the new revision IDs.
```

Author: Great point.

## Safe Expunging Process

> SCSs MAY allow the organization to expunge (remove) content from a repository and its change history without leaving a public record of the removed content.

Reviewer 4: We could add a sentence here to help set the tone. 

```
Removing a revision from a repository is similar to deleting a package version from a registry: it's almost impossible to estimate the amount of downstream supply chain impact. 

In version control systems like git, removal of a revision changes the object id of all subsequent revisions that were built on top of it. 

Although there is no "safe" way to do it, it sometimes necessary and there are steps you can take to mitigate the damage.

Author: Done.

This includes changing files, history, or changing references in git and is used to accommodate legal/privacy compliance requirements as well as administrative

> changes within a repository (see below for more information on the various scenarios).

Author: I think this is resolved now as we've removed this section?

When used as an attack, this is called “repo hijacking” (or “repo-jacking”) and is one of the primary threats source provenance attestations protect against.
As such great care must be taken by SCSs when implementing these changes for legitimate purposes.

SCSs SHOULD require multi-party approval for any changes made under this process.

The Safe Expunging Process MUST be documented and describe how requests and actions are tracked and SHOULD log the fact that content was removed.

On the git VCS, force pushes allow you to remove data from a branch.
If a branch has been identified as consumable branch, force pushes to that branch must follow the safe expunging process.

Different organizations and tech stacks may have different approaches to the problem.

Safe Expunging Scenarios (non-exhaustive):

### Legal Takedowns

> A legal takedown request will be addressed by following an agreed-upon process.

Author: I suspect it could be too restrictive.  We've seen some examples in National Security Letters [that forbid disclosure](https://en.wikipedia.org/wiki/National_security_letter#:~:text=NSLs%20typically%20contain%20a%20nondisclosure%20requirement%20forbidding%20the%20recipient%20of%20an%20NSL%20from%20disclosing%20the%20FBI%20had%20requested%20the%20information).  I could see something like that happening here too.

I figure laws are written by governments which have their own agendas and may not care what SLSA says.

and regulations which may require the change to be kept private to the extent possible.

### Commit metadata rewrites

A team may decide that all reachable commits in the history of a revision need to follow a new metadata convention.
In git VCS, compliance with this new policy will require history to be rewritten (commit metadata is included in the computation of the revision id).
Policies in this category include things like commit signatures, author / committer formatting restrictions, closed-issue-linkage, etc.

Organizations MUST make the fact of these changes public when they occur (e.g. an issue, announcement, or other mechanism).

### Repository renames

When a repo is transferred to a new organization ("donated"), or if a repo must be renamed or otherwise have its url changed within the same org, attestations for previous revisions of this repo will no longer be matched because the combination of the repository id and the revision id will have changed.

Organizations MUST inform all consumers of a repo of these changes when they occur.

> Organizations SHOULD provide a means to resolve previous revision IDs to the new revision IDs.

Author: Removed this section entirely.
